notes/hashfunctions3.tex

```diff
@@ -0,0 +1,353 @@
+            \myInd $\ftable[X] \getsr \bits^n$\\
+            Ret $\ftable[X]$
+        }
+        {
+            \underline{$\INDIFF0^\advD_{\Horacle, \simoracle}$}\\[1pt]
+            $b' \getsr \advD^{\FnOracle, \simoracle}$\\
+            Ret $b'$\medskip
+        
+            \underline{$\FnOracle(M)$}\\
+            If $\Htable[M] = \bot$ then\\
+            \myInd $\Htable[M] \getsr \bits^n$\\
+            Ret $\Htable[M]$\\
+            
+            \underline{$\simoracle(X)$}\\
+            Ret $\mathcal{S}^\FnOracle[X]$
+        }
+    }
+    \caption{Indifferentiability from a Random Oracle} \label{fig:Indiff}
+\end{figure}
+
+\noindent We define a distinguisher $\advD$'s advantage in the $\INDIFF$ game as:
+\[
+\AdvINDIFF{\hash, f, \mathcal{S}}{\advD} = 
+\left| 
+    \Pr[\INDIFF1^\advD_{\hash,f} \Rightarrow 1] 
+    - \Pr[\INDIFF0^\advD_{\Horacle,\mathcal{S}} \Rightarrow 1]
+\right|
+\]
+
+\begin{wrapfigure}{r}{1.5in}
+    \centering
+    \fpage{.2}{
+            \underline{$\textbf{adversary } \advD^{\FnOracle, O}$}\\[1pt]
+            $m_1, m_2 \getsr \bits^d$\\
+            $y_1 \gets \FnOracle(m_1)$\\
+            $y_2 \gets O(y_1,  m_2)$\\
+            $y'_2 \gets \FnOracle(m_1 \Vert m_2)$\\
+            Ret $(y_2 = y'_2)$
+    }
+    \caption{MD adversary} \label{fig:Indiff distinguisher}
+\end{wrapfigure}
+
+\noindent First, we note that a good simulator $\simoracle$ for $f$ in the ideal world may not always exist. In fact, the Merkle Damg{\aa}rd construction is actually not indifferentiable from a random oracle. Consider the $\INDIFF$-adversary $\advD$ defined in Figure~\ref{fig:Indiff distinguisher}.
+In the real world, the two outputs $y_2$ and $y'_2$ will always be equal. On the other hand, in the ideal world, the probability that $y_2$ and $y'_2$ are equal is just the probability that a randomly sampled $n$ bit string is equal to the $y_2$ output by the simulator; this is equal to $\frac{1}{2^n}$. Therefore, $\advD$ distinguishes between the real and ideal worlds with high probability.
+
+Variations of the Merkle Damg{\aa}rd construction such as the Chopped MD or the Enveloped MD are often used in practice instead since they are indifferentiable from a random oracle
+
+\bigskip
+\noindent How is this indifferentiable framework useful? It's real power is seen through the following composition theorem which allows security analysis of a construction that uses the hash function $\hash^f$ for any arbitrary game $G$
+
+\begin{theorem}
+\label{thm: indiff composition}
+\textbf{Indifferentiability Composition Theorem} \cite{Maurer2004}
+For any security game $G$, construction $\hash^f$ and scheme $C$, if 
+\begin{enumerate}
+    \item $\hash^f$ is indifferentiable from a random oracle under the assumption that $f$ is a random oracle
+    \item C is provably $G$-secure in the random oracle model
+\end{enumerate}
+then, $C$ is provably $G$-secure using $\hash^f$ under the assumption that $f$ is a random oracle
```

Comment:
I know this is how Tom stated the thm in the lecture slides, but might be more formal to specify what G-secure and indifferentiable mean by stating the theorem in terms of an advantage bound. Something like:
Pr[G^{\advA, RO} =>1] \leq Pr[G^{\advA, \hash}  => 1] + \AdvIndiff^{\advB}

Reply:
Can check with Tom to see if that makes sense...